module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,764 @@
+            Ok(CallResponse::default())
+        } else {
+            Err(self.slash_burn_reward(
+                attester,
+                Role::Attester,
+                SlashingReason::TransitionNotFound,
+                working_set,
+            ))
+        }
+    }
+
+    fn check_initial_hash(
+        &self,
+        max_attested_height: u64,
+        attester: &C::Address,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // Normal state
+        if let Some(transition) = self
+            .chain_state
+            .historical_transitions
+            .get(&(max_attested_height - 1), working_set)
+        {
+            if transition.post_state_root() != attestation.initial_state_root {
+                // The initial root hashes don't match, just slash the attester
+                return Err(self.slash_burn_reward(
+                    attester,
+                    Role::Attester,
+                    SlashingReason::InvalidInitialHash,
+                    working_set,
+                ));
+            }
+        } else {
+            // Genesis state
+            // We can assume that the genesis hash is always set, otherwise we need to panic.
+            // We don't need to prove that the attester was bonded, simply need to check that the current bond is higher than the
+            // minimal bond and that the attester is not unbonding
+            if self
+                .chain_state
+                .get_genesis_hash(working_set)
+                .expect("The initial hash should be set")
+                != attestation.initial_state_root
+            {
+                // Slash the attester, and burn the fees
+                return Err(self.slash_burn_reward(
+                    attester,
+                    Role::Attester,
+                    SlashingReason::InvalidInitialHash,
+                    working_set,
+                ));
+            }
+            // Normal state
+        }
+
+        Ok(CallResponse::default())
+    }
+
+    /// Try to process an attestation, if the attester is bonded
+    pub(crate) fn process_attestation(
+        &self,
+        context: &C,
+        attestation: Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // We first need to check that the attester is still in the bonding set
+        if self
+            .bonded_attesters
+            .get(context.sender(), working_set)
+            .is_none()
+        {
+            return Err(AttesterIncentiveErrors::UserNotBonded);
+        }
+
+        // We first need to check the bonding proof (because we can't slash the attester if he's not bonded)
+        self.check_bonding_proof(context, &attestation, working_set)?;
```

Comment:
This check looks like it has already been handled by ` if self.bonded_attesters.get(context.sender(), working_set).is_none() { return Err(_)}`. I think there is an edge that we need to handle where the attester is bonded, but he provides an invalid proof of bond - is that what this check is intended to do? If so, we can just change the comment to say "If the bonding proof in the attestation is invalid, light clients will ignore the attestation. In that case, we should too. "

Reply:
Yes that's exactly what this is meant to do, I updated the code after your review but did not change the comment!